### PR TITLE
tint in ParticleContainer

### LINKED ISSUE #3777: ParticleContainer.js should support tint optionality, as many effects require this
Doesn't seem like it would effect performance all that much, unless i'm missing something. The tint is applied in shader correct, and a CPU calculation done in client to invert the hex.

... Specifically, it becomes possible to animate the tint, or re-use the same "glow" texture in many different situations, especially if your animation data is pre-rendered.

### PR DESCRIPTION
**Pull request by ivanpopelyshev**
Trying to fix #3777

### CODE REVIEW

File: src/particles/ParticleContainer.js [hunk 1]
```diff
@@ -149,7 +149,11 @@ export default class ParticleContainer extends core.Container
             this._properties[1] = 'position' in properties ? !!properties.position : this._properties[1];
             this._properties[2] = 'rotation' in properties ? !!properties.rotation : this._properties[2];
             this._properties[3] = 'uvs' in properties ? !!properties.uvs : this._properties[3];
-            this._properties[4] = 'alpha' in properties ? !!properties.alpha : this._properties[4];
+            this._properties[4] = ('alpha' in properties) ? !!properties.alpha : this._properties[4];
```
Comment (englercj) [inline src/particles/ParticleContainer.js]: unnecessary change?

File: src/particles/ParticleContainer.js [hunk 2]
```diff
@@ -149,7 +149,11 @@ export default class ParticleContainer extends core.Container
...
+            if ('tint' in properties)
+            {
+                this._properties[4] = !!properties.tint;
+            }
```
Comment (englercj) [inline src/particles/ParticleContainer.js]: I don't think this logic is quite right, if I set `alpha` to true, but `tint` to false then alpha doesn't get uploaded. Probably need to change this to something like:

```js
this._properties[4] = 'alpha' in properties || 'tint' in properties ? !!properties.alpha || !!properties.tint : this._properties[4];
```

That way either of them being set will use the value, and either of them being set to true will cause the upload.

File: src/particles/webgl/ParticleRenderer.js [hunk 3]
```diff
@@ -387,16 +389,21 @@ export default class ParticleRenderer extends core.ObjectRenderer
...
+            const premultiplied = sprite._texture.baseTexture.premultipliedAlpha;
+            const alpha = sprite.alpha;
+            // we dont call extra function if alpha is 1.0, that's faster
+            const argb = alpha < 1.0 && premultiplied ? premultiplyTint(sprite._tintRGB, alpha)
+                : sprite._tintRGB + (alpha * 255 << 24);
+
+            array[offset] = argb;
+            array[offset + stride] = argb;
+            array[offset + (stride * 2)] = argb;
+            array[offset + (stride * 3)] = argb;
```
Comment (englercj) [inline src/particles/webgl/ParticleRenderer.js]: Looking this and the other functions, I'm not clear why there are 4 assignments in them? If the stride is 1, and you are uploading UNSIGNED_BYTE to a vec4, wouldn't you only assign `array[offset]`?

Why do you have to fill 4 Uint32s for an UNSIGNED_BYTE vec4? This pattern is used in all the upload funcs, just not sure why each of them copy the data 4 times?

Comment (ivanpopelyshev) [reply src/particles/webgl/ParticleRenderer.js]: That's how our SpriteRenderer works. 4 assignments are for different vertices, and its not clear what stride is, its always more than zero. This abstraction exists because we dont know which fields are static and which are dynamic.

File: src/particles/ParticleContainer.js (hunk 2 above)
Comment (ivanpopelyshev) [reply src/particles/ParticleContainer.js]: Agree.

File: src/particles/webgl/ParticleRenderer.js (hunk 3 above)
Comment (englercj) [reply src/particles/webgl/ParticleRenderer.js]: Ah, that's right we upload it per vert (one for each uv coord). Forgot about that! Thanks!

**Comment by ivanpopelyshev** [comment]
I didnt think it would be approved that fast :)

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev is it possible to see a jsfiddle example of this functionality?

**Comment by ivanpopelyshev** [comment]
@bigtimebuddy it is possible in pixi-examples, just run https://pixijs.github.io/examples/#/basics/basic.js with current version and with new one. Suddenly, tint starts working!

**Comment by bigtimebuddy** [comment]
Verified! Great work @ivanpopelyshev, you're a machine
